### DISCUSSION TITLE
IfcRelAggregates entity duplication?

### DISCUSSION THREAD

**Discussion by karfous** [IfcOpenShell/IfcOpenShell#5861] [Q&A] [answered]
Hello community, I have just started experimenting with IfcRelAggregates and I have a simple hello world file. While opening this file BIM vision is telling me that there are some duplicate elements. This is caused by #176 but why? I am just referencing this relations to #159 and #131 . Can you explain please? 
![image](https://github.com/user-attachments/assets/c8398c1f-a256-4446-9405-e80e7b49ad73)

`ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition [Ifc4X3NotAssigned]'),'2;1');
FILE_NAME(
/* name */ 'C:\\Users\\KaresJak\\Desktop\\cubes\\cubes.ifc',
/* time_stamp */ '2024-12-09T15:02:43',
/* author */ ('KaresJak'),
/* organization */ ('Unknown'),
/* preprocessor_version */ 'GeometryGymIFCcore v0.2.19.0 by Geometry Gym Pty Ltd built 2023-09-24T19:25:26',
/* originating_system */ 'Civil 3D 2024 IfcInfra Plugin v13.6.147.0',
/* authorization */ 'None'
);

FILE_SCHEMA (('IFC4X3_ADD2'));
ENDSEC;

DATA;
#1= IFCORGANIZATION($,'Autodesk',$,$,$);
#2= IFCAPPLICATION(#1,'Civil 3D 2024','Civil 3D','Civil 3D 2024 IfcInfra Plugin v13.6.147.0');
#3= IFCCARTESIANPOINT((0.0,0.0,0.0));
...
#37= IFCPRODUCTDEFINITIONSHAPE($,$,(#36));
#38= IFCANNOTATION('1gyIk2JnP5XQASinIxMmzO',$,'IFC Project Base Point','IFC Project Base Point','CogoPoint',#28,#37,$);
#39= IFCRELCONTAINEDINSPATIALSTRUCTURE('2X3O97QQ9tIHlRr7wiHYsC',$,$,' Container for Elements',(#38,#131,#159),#9);
#40= IFCPROPERTYSINGLEVALUE('PointNumber',$,IFCIDENTIFIER('1'),$);
#41= IFCPROPERTYSET('2V_rtKmpT37BRP4TEuir80',$,'C3D_Attributes',$,(#40));
#42= IFCRELDEFINESBYPROPERTIES('1SHKyJmUb8SAvPIC9ZLyjT',$,$,$,(#38),#41);
...
#130= IFCLOCALPLACEMENT(#5,#27);
#131= IFCBUILTELEMENT('1iihuXPuj4nQEql2cVRn3L',$,'0',$,$,#130,#129,'DB3');
#132= IFCSTYLEDITEM(#127,(#45),'0');
#133= IFCPROPERTYSET('0_ukj2eHL4C96ep7BUC4vR',$,'Properties',$,(#80,#81,#82,#83,#84,#85,#86,#87,#88,#89,#90,#91,#92,#93,#94,#95,#96,#97,#98,#100,#101,#135,#136,#137,#138,#139,#140,#141));
#134= IFCRELDEFINESBYPROPERTIES('2eEnsrDWb7nB8HCk6gq1TA',$,$,$,(#131),#133);
#135= IFCPROPERTYSINGLEVALUE('Area',$,IFCREAL(6.0),$);
#136= IFCPROPERTYSINGLEVALUE('IsNull',$,IFCBOOLEAN(.F.),$);
...
#157= IFCPRODUCTDEFINITIONSHAPE($,$,(#156));
#158= IFCLOCALPLACEMENT(#5,#27);
#159= IFCBUILTELEMENT('2QxNdW4fD5mgXI$uijawiv',$,'0',$,$,#158,#157,'DB7');
#160= IFCSTYLEDITEM(#155,(#45),'0');
#161= IFCPROPERTYSET('1vaboyA3f5COaK86JKTR6r',$,'Properties',$,(#80,#81,#82,#83,#84,#85,#86,#87,#88,#89,#90,#91,#92,#93,#94,#95,#96,#97,#98,#100,#101,#135,#136,#137,#138,#139,#140,#163));
#162= IFCRELDEFINESBYPROPERTIES('3DznxKAMD3GvtEaDfoWwqF',$,$,$,(#159),#161);
...
#176 = IFCRELAGGREGATES('5Fv6DjW8rfRf7aVcXpR6Df',$,'Assembly Aggregation',$,#159,(#131));

ENDSEC;

END-ISO-10303-21;
`

**Comment by aothms** [comment, marked as answer]
Because `#131` has both the incoming aggregation relationship as well as the contained in structure relationship it is essentially is contained in the decomposition tree twice.

You can see the diagram here https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcRelVoidsElement.htm Although that is about voiding, the same applies to aggregation, do not add containment relationships to the aggregation children

**Reply by karfous** [reply]
Thank you, I need to get used to ifc documentation. It is a little bit complicated in comparison with javascript MDN docs from Mozilla :D ... thank you again I understand now - the error is on #39 - #131 should not be mentioned there
Wrong
`#39= IFCRELCONTAINEDINSPATIALSTRUCTURE('2X3O97QQ9tIHlRr7wiHYsC',$,$,' Container for Elements',(#38,#131,#159),#9);`
Correct
`#39= IFCRELCONTAINEDINSPATIALSTRUCTURE('2X3O97QQ9tIHlRr7wiHYsC',$,$,' Container for Elements',(#38,#159),#9);`
